fedimint-client/src/lib.rs

```diff
@@ -1552,18 +1761,42 @@ impl ClientBuilder {
         Ok(client)
     }
 
+    pub async fn join(
```

Comment:
Needs documentation on when to use imo (with some big warnings about risk of fund loss if re-joining a federation with a previously used key without recovering). Even if the user never provided the key and it was randomly generated, but e.g. an app allowed leaving+joining federations that will be a problem.

Maybe we should always call recover and only attempt recovery if there is a backup (since otherwise it will take forever for real federations anyway)?

Reply:
Oh, I actually wanted to have a top level secret key to be an `enum`:

```rust
enum RootSecret {
  Fresh(DeriveableSecret),
  FromUser(DeriveableSecret),
}
```

so that `Fresh` is only returned when the key is randomly generated, and `FromUser` when entered by the user. This way type system would steer the implementation to handle it correctly, and it would be a good place to put docs explaining why it's needed.

Reply:
Ehhh... I wrote some docstrings, but it made me realize that without some reliable way to tell if a given `root_secret` ever joined Federations, **most** apps will have to pessimize and call `recover` a lot which is particularly slow when a given Federation was not used (as it causes a whole history scan, because there can't be any previous backup).

> Maybe we should always call recover and only attempt recovery if there is a backup (since otherwise it will take forever for real federations anyway)?

And I'm not sure if relying on a backup is fundamentally sound. I wouldn't be surprised if backups will be a source of DoS and popular Federations will need to disable it altogether, or prune very old ones etc.